pulp_rpm/app/serializers/repository.py

```diff
@@ -308,6 +309,29 @@ class RpmRepositorySyncURLSerializer(RepositorySyncURLSerializer):
     Serializer for RPM Sync.
     """
 
+    mirror = serializers.BooleanField(
+        required=False,
+        allow_null=True,
+        help_text=_(
+            "DEPRECATED: If ``True``, ``sync_policy`` will default to 'mirror_complete' "
+            "instead of 'additive'."
+        ),
+    )
+    sync_policy = serializers.ChoiceField(
```

Comment:
we probably should take into account the policy change in between the syncs which should affect whether the sync is opimized or not. Otherwise once i sync with additive and then change to mirror_complete I don't seem to have any option how to mirror the metadata and have publication created.
```
$ http :24817/pulp/api/v3/repositories/rpm/rpm/d45f7f90-bfd2-43d9-8832-aa51f52a30c0/sync/ remote=/pulp/api/v3/remotes/rpm/rpm/3a95a54f-31cb-40ff-a7fa-33c38ed8085d/ sync_policy=additive
HTTP/1.1 202 Accepted
Access-Control-Expose-Headers: Correlation-ID
Allow: POST, OPTIONS
Connection: close
Content-Length: 67
Content-Type: application/json
Correlation-ID: d319a4d4a0f74cf1bdd5961711ab0724
Date: Fri, 15 Oct 2021 10:54:55 GMT
Referrer-Policy: same-origin
Server: gunicorn
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "task": "/pulp/api/v3/tasks/9611807e-4744-44ab-9b5e-7a09374bd755/"
}


(pulp) [vagrant@pulp3-source-fedora34 _scripts]$ http :24817/pulp/api/v3/repositories/rpm/rpm/d45f7f90-bfd2-43d9-8832-aa51f52a30c0/sync/ remote=/pulp/api/v3/remotes/rpm/rpm/3a95a54f-31cb-40ff-a7fa-33c38ed8085d/ sync_policy=mirror_complete
HTTP/1.1 202 Accepted
Access-Control-Expose-Headers: Correlation-ID
Allow: POST, OPTIONS
Connection: close
Content-Length: 67
Content-Type: application/json
Correlation-ID: 38a39f7b161e4b2aa860d4c47e69773d
Date: Fri, 15 Oct 2021 10:55:05 GMT
Referrer-Policy: same-origin
Server: gunicorn
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "task": "/pulp/api/v3/tasks/d61228b7-4d2f-47eb-ac05-2e7c1723d630/"
}


(pulp) [vagrant@pulp3-source-fedora34 _scripts]$ http :24817/pulp/api/v3/tasks/d61228b7-4d2f-47eb-ac05-2e7c1723d630/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, PATCH, DELETE, HEAD, OPTIONS
Connection: close
Content-Length: 829
Content-Type: application/json
Correlation-ID: 098bcf4bc5f14ded8434e2ee4eb16e25
Date: Fri, 15 Oct 2021 10:55:13 GMT
Referrer-Policy: same-origin
Server: gunicorn
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "child_tasks": [],
    "created_resources": [],
    "error": null,
    "finished_at": "2021-10-15T10:55:08.356237Z",
    "logging_cid": "38a39f7b161e4b2aa860d4c47e69773d",
    "name": "pulp_rpm.app.tasks.synchronizing.synchronize",
    "parent_task": null,
    "progress_reports": [
        {
            "code": "sync.was_skipped",
            "done": 1,
            "message": "Skipping Sync (no change from previous sync)",
            "state": "completed",
            "suffix": null,
            "total": null
        }
    ],
    "pulp_created": "2021-10-15T10:55:05.831809Z",
    "pulp_href": "/pulp/api/v3/tasks/d61228b7-4d2f-47eb-ac05-2e7c1723d630/",
    "reserved_resources_record": [
        "/pulp/api/v3/repositories/rpm/rpm/d45f7f90-bfd2-43d9-8832-aa51f52a30c0/",
        "shared:/pulp/api/v3/remotes/rpm/rpm/3a95a54f-31cb-40ff-a7fa-33c38ed8085d/"
    ],
    "started_at": "2021-10-15T10:55:05.908091Z",
    "state": "completed",
    "task_group": null,
    "worker": "/pulp/api/v3/workers/15291194-b623-4182-a0c9-a43056e75bef/"
}


(pulp) [vagrant@pulp3-source-fedora34 _scripts]$ http :24817/pulp/api/v3/repositories/rpm/rpm/d45f7f90-bfd2-43d9-8832-aa51f52a30c0/sync/ remote=/pulp/api/v3/remotes/rpm/rpm/3a95a54f-31cb-40ff-a7fa-33c38ed8085d/ sync_policy=mirror_complete optimize=false
HTTP/1.1 202 Accepted
Access-Control-Expose-Headers: Correlation-ID
Allow: POST, OPTIONS
Connection: close
Content-Length: 67
Content-Type: application/json
Correlation-ID: 4841e3b5ff1b4705bf71ba4046d09d52
Date: Fri, 15 Oct 2021 10:55:20 GMT
Referrer-Policy: same-origin
Server: gunicorn
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "task": "/pulp/api/v3/tasks/f934247a-a748-4119-85e7-20a27ee6afae/"
}


(pulp) [vagrant@pulp3-source-fedora34 _scripts]$ http :24817/pulp/api/v3/tasks/f934247a-a748-4119-85e7-20a27ee6afae/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, PATCH, DELETE, HEAD, OPTIONS
Connection: close
Content-Length: 1536
Content-Type: application/json
Correlation-ID: 4fdd010d6c6f46f2942f0e9cd0707edc
Date: Fri, 15 Oct 2021 10:55:31 GMT
Referrer-Policy: same-origin
Server: gunicorn
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "child_tasks": [],
    "created_resources": [],
    "error": null,
    "finished_at": "2021-10-15T10:55:29.510888Z",
    "logging_cid": "4841e3b5ff1b4705bf71ba4046d09d52",
    "name": "pulp_rpm.app.tasks.synchronizing.synchronize",
    "parent_task": null,
    "progress_reports": [
        {
            "code": "sync.parsing.advisories",
            "done": 4,
            "message": "Parsed Advisories",
            "state": "completed",
            "suffix": null,
            "total": 4
        },
        {
            "code": "sync.downloading.metadata",
            "done": 10,
            "message": "Downloading Metadata Files",
            "state": "completed",
            "suffix": null,
            "total": null
        },
        {
            "code": "sync.parsing.packages",
            "done": 35,
            "message": "Parsed Packages",
            "state": "completed",
            "suffix": null,
            "total": null
        },
        {
            "code": "sync.parsing.comps",
            "done": 3,
            "message": "Parsed Comps",
            "state": "completed",
            "suffix": null,
            "total": 3
        },
        {
            "code": "unassociating.content",
            "done": 0,
            "message": "Un-Associating Content",
            "state": "completed",
            "suffix": null,
            "total": null
        },
        {
            "code": "sync.downloading.artifacts",
            "done": 0,
            "message": "Downloading Artifacts",
            "state": "completed",
            "suffix": null,
            "total": null
        },
        {
            "code": "associating.content",
            "done": 0,
            "message": "Associating Content",
            "state": "completed",
            "suffix": null,
            "total": null
        }
    ],
    "pulp_created": "2021-10-15T10:55:20.554636Z",
    "pulp_href": "/pulp/api/v3/tasks/f934247a-a748-4119-85e7-20a27ee6afae/",
    "reserved_resources_record": [
        "/pulp/api/v3/repositories/rpm/rpm/d45f7f90-bfd2-43d9-8832-aa51f52a30c0/",
        "shared:/pulp/api/v3/remotes/rpm/rpm/3a95a54f-31cb-40ff-a7fa-33c38ed8085d/"
    ],
    "started_at": "2021-10-15T10:55:20.620772Z",
    "state": "completed",
    "task_group": null,
    "worker": "/pulp/api/v3/workers/15291194-b623-4182-a0c9-a43056e75bef/"
}


(pulp) [vagrant@pulp3-source-fedora34 _scripts]$

Reply:
one more thing crossed my mind.. should not autopublish be incompatible with mirror_complete?

Reply:
This observation is unrelated to this pr but how retain_package_version changes would come in place in the following scenario:
1. create, sync repo  in additive more
2. update repo to keep just 1 packge version
3. sync repo again but it did not change remotely

Reply:
@ipanova I'm actually covering this in my other PR

https://github.com/pulp/pulp_rpm/pull/2148/commits/10e15d223511d491197829161c77c43962416f1c#diff-fb406f31027edc1c5b65abe7635bbd2830a7cc05c24a487344d7649017d14628R322